20240129-satellite-managed-encryption.md

```diff
@@ -0,0 +1,188 @@
+* the project ID of the project that the Secret Manager is in.
+* the secret ID of the master key, which at this point should already be created in the Secret Manager.
+
+Alternatively, we can use the [GCP Secret Store CSI driver](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp) to mount the Master Key in the Satellite pod without having to expose any credentials at all.
+However, there are [security considerations](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp?tab=readme-ov-file#security-considerations) we have to make.
+
+#### Satellite Project Updates
+
+A new column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and `null` for others.
+Projects that are created before this feature is deployed and projects created after but opt out should have `null` in this column.
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the KMS service generates a cryptographically random passphrase and encrypts it using the master key
+* the passphrase store is updated with the encrypted passphrase for this project
+* endpoints on the satellite API are added or updated so that an authenticated user can request the encryption passphrase for a project during access creation
+  * endpoints handlers should use the KMS service to retrieve the encrypted passphrase from the store and decrypt it using the master key
+
+#### Satellite UI Updates
+
+When creating a project, the user should be provided the option to create different types of projects:
+
+* satellite-managed-encryption project
+* user-managed-encryption project
+
+The user-managed-encryption project will be the same as before, and user will always be prompted to enter encryption passphrase when necessary, e.g. when using the file browser or generating an access grant.
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
+
+The user should not be able to directly retrieve a "raw API key" from the Satellite UI. This prevents the user from unintentionally creating "invalid" credentials using a raw API key with Uplink. The only credentials that can be generated from the access page for a satellite-managed-encryption project should be credentials that already have embedded encryption rules.
+
+The file browser in the UI should generate access credentials using the passphrase from the satellite KMS, and no path encryption.
```

Comment:
We could, but is it necessary? If the UI is designed to always add the necessary restrictions (encryption rules) to any raw API key before providing it to the user.
Maybe a user could still get the raw key and modify the encryption info if they knew what they were doing, but it would not be possible to do accidentally, I think.
What is the advantage of adding a new DRPC service in your opinion?